Adding go_v2 - a more idiomatic Go implementation.

EscapeIndustries opened the pull request:
This is my Go implementation of your Pi server identifier problem.

Please note I've put this code under the MIT license by including it in comments at the head of my file. I find the GPLv3 obnoxious.

EscapeIndustries committed:
Adding go_v2 - a more idiomatic Go implementation.

alexellis commented on go_v2/main.go:
```diff
@@ -0,0 +1,84 @@
+/*
```
If we change the repo to a MIT license, can we take it out of the file as an include?

EscapeIndustries replied:
It would be nice to switch it all to MIT, but you still have to have _something_ in each file to claim copyright and refer to the license. The MIT is short so I took the path of least resistance. Would you like a PR where I either add the license to every file, or change the top level license and add an appropriate reference to each file?

alexellis replied:
I've altered the origin repo to have a MIT license at the root, not sure if I've done it correctly. Top level plus reference to each file I think would be a bit leaner. 

You may need to pull from upstream first if you want to update anything I've changed since your fork.

alexellis commented on go_v2/main.go:
```diff
@@ -0,0 +1,84 @@
+	go printUnique(hexOut)
+	var wg sync.WaitGroup
+	wg.Add(requestLimit)
+	for i := 0; i < requestLimit; i++ {
```
Being a Go novice.. is there any throttling on the total amount of requests open at one time?

EscapeIndustries replied:
There is no throttling so this version of the code spins up all the requests in parallel. Adding throttling won't add very much code - I was just following the super-minimal pattern set by the other implementations.

I'll add a limit to work-in-progress and you can see how little extra code it adds.

alexellis replied:
I think it's good to have the minimal implementation, how about two go versions where one also has throttling? Finding this very interesting.

alexellis commented:
@EscapeIndustries thanks for the more idiomatic Go implementation. Are you also OK to add a brief README.md in the go_v2/folder?

EscapeIndustries committed:
Introduce concurrency limits and begin a README.md file.

EscapeIndustries committed:
Catch up with upstream after Alex switched the license from GPLv3 to MIT.

EscapeIndustries committed:
Added usage to README

EscapeIndustries committed:
Limit concurrency to a pool of workers.

EscapeIndustries committed:
Fix grammar in Go v2 README

EscapeIndustries committed:
Fix grammar in README.

EscapeIndustries committed:
Tidy up copyright and license notices

EscapeIndustries committed:
More compact copyright and license comments

EscapeIndustries commented:
At this point I have added concurrency limits, added a readme and amended the copyright/license notices.
The extra code for the concurrency limit is so little that I'm not sure a separate version makes sense. Take a look?

The one thing I might still do is put a bunch of comments in the code file explaining how it does what it does, plus optional logging to visualise the program flow. BUT this isn't a language tutorial, so that is probably overkill.

alexellis commented:
@EscapeIndustries A few brief very comments would probably be useful (for my own Go journey), I'm going to ask you a bit about the work limiting when I see you at the office. Thanks for this great contribution 👍
